x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStopTransformAction.java

```diff
@@ -153,13 +153,24 @@ protected void taskOperation(Request request, TransformTask transformTask, Actio
         }
 
         if (ids.contains(transformTask.getTransformId())) {
-            try {
-                transformTask.stop(request.isForce());
-            } catch (ElasticsearchException ex) {
-                listener.onFailure(ex);
-                return;
-            }
-            listener.onResponse(new Response(Boolean.TRUE));
+            transformTask.setShouldStopAtCheckpoint(request.isWaitForCheckpoint(), ActionListener.wrap(
+                r -> {
+                    try {
+                        transformTask.stop(request.isForce(), request.isWaitForCheckpoint());
+                        listener.onResponse(new Response(true));
+                    } catch (ElasticsearchException ex) {
+                        listener.onFailure(ex);
```

Comment:
nit: as this causes often trouble, an explicit return would be good although superfluous logically here.